### ISSUE TITLE
dart2js compiler issue with generics around Dart 2.8.0-dev.1

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/sdk#40296]
See https://travis-ci.org/FirebaseExtended/firebase-dart/jobs/638163526

2.8.0-dev.0.0 – works fine
2.8.0-dev.1.0 – test runner hangs compiling
2.8.0-dev.2.0 – start getting test failures, mostly due to timeouts
2.8.0-dev.5.0 – still failing

CC @sigmundch @fishythefish

**Comment by fishythefish** [comment]
To clarify, 2.8.0-dev.1.0 wasn't actually hanging during compilation. Rather, we hit a bad assertion (fixed by aaaaa479f3113a2727d9cbe8a02136faa7551c14) while the test was still being loaded. This left the communication channel in a half-open state. In fact, the compilation was successful, but the loading message just says that the test is compiling when, in reality, it's waiting for the test to be loaded.
(cc @jakemac53)

2.8.0-dev.2.0 included the fix for the test loading issue, which uncovered the actual issue. Presumably, the issue was either introduced in -dev.1.0 and masked by the test loading issue or introduced in -dev.2.0 along with the test loading fix.

To further bisect, try running -dev.1.0 without enabling asserts.

**Comment by kevmoo** [comment]
It does seem to happen in `2.8.0-dev.1.0` (when I manually disable asserts in the test runner)

```
TypeError: Instance of 'gd': type 'UnknownJavaScriptObject' is not a subtype of type 'FutureOr<QuerySnapshotJsImpl>'
```

**Comment by sigmundch** [comment]
Small repro:
**a.js**:
```js
function A() {
  this.x = 1;
}
self.foo = new A();
```

**b.dart**:
```dart
import 'dart:async';
import 'dart:html';
import 'package:js/js.dart';

@JS('A')
@anonymous
class MyFoo {}

@JS('foo')
external dynamic get foo;

@pragma('dart2js:noInline')
check<T>() {
  FutureOr<T> x = foo;
}

main() {
  check<MyFoo>(); // causes a check for UnknownJavaScriptObject is FutureOr<MyFoo>
  window.postMessage(null, null); // implies adding JSObject as a supertype of UnknownJavaScriptObject
}
```

**Comment by kevmoo** [comment]
@fishythefish – confirmed this fixes the failing test!

**Comment by kevmoo** [comment]
Fix landed in `2.8.0-dev.6.0` – confirmed fix in firebase package!
